File: SuperGrid/VB/DemoBasicFiltering.vb
Imports System.ComponentModel
Imports DevComponents.DotNetBar.SuperGrid

Namespace SuperGridDemo
	Partial Public Class DemoBasicFiltering
        Inherits DevComponents.DotNetBar.Office2007Form
		#Region "Private variables"

		Private _Rand As New Random()

		#End Region

		Public Sub New()
			InitializeComponent()

			' Initialize the grid
			' and set the sample description text

			InitializeGrid()

            ShellServices.LoadRtbText(richTextBox1, "SuperGridDemo.DemoBasicFiltering.rtf")
		End Sub

		#Region "InitializeGrid"

		''' <summary>
		''' Initialize the grid
		''' </summary>
		Private Sub InitializeGrid()
			Dim panel As GridPanel = superGridControl1.PrimaryGrid

			panel.Columns("LastName").EditorType = GetType(MyComboBox)
			panel.Columns("Citizen").EditorType = GetType(MyCheckBox)

			' Add some initial rows for the user to filter

			AddRows()
		End Sub

		#End Region

		#Region "AddRows"

		''' <summary>
		''' Adds 1000 random data rows for the user to filter
		''' </summary>
		Private Sub AddRows()
			Dim panel As GridPanel = superGridControl1.PrimaryGrid

			superGridControl1.BeginUpdate()

			Dim count As Integer = panel.Rows.Count

			For i As Integer = 0 To 999
				Dim emp As MyEmployee = MyEmployee.GetNewEmployee()

				Dim firstName As Object
				If emp.Age Mod 7 = 0 Then
					firstName = (CObj(DBNull.Value))
				Else
					firstName = (emp.FirstName)
				End If
				Dim age As Object
				If emp.Age Mod 5 = 0 Then
					age = (Nothing)
				Else
					age = (CObj(emp.Age))
				End If
				Dim citizen As Object
				If emp.Age Mod 3 = 0 Then
					citizen = (Nothing)
				Else
					citizen = (CObj(emp.Citizen))
				End If
				Dim hireDate As Object
				If emp.Age Mod 6 = 0 Then
					hireDate = (Nothing)
				Else
					hireDate = (CObj(emp.HireDate))
				End If

				Dim ob1() As Object = { count + i, emp.LastName, firstName, age, hireDate, citizen }

				If i < 12 Then
					Dim n As Integer = i Mod 6 + 1

					Dim ob2(n - 1) As Object

					For j As Integer = 0 To n - 1
						ob2(j) = ob1(j)
					Next j

					panel.Rows.Add(New GridRow(ob2))
				Else
					panel.Rows.Add(New GridRow(ob1))
				End If
			Next i

			superGridControl1.EndUpdate()

			' We'll stop it at 10,000 - just for grins

			If panel.Rows.Count >= 10000 Then
				btnAddRows.Enabled = False
			End If
		End Sub

		#End Region

		#Region "BtnAddRowsClick"

		''' <summary>
		''' Handles user request to add another 1000 rows
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnAddRowsClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnAddRows.Click
			AddRows()
		End Sub

		#End Region

		#Region "BtnResetPanelFilterClick"

		''' <summary>
		''' Handles user request to reset the
		''' row level panel filter
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnResetPanelFilterClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnResetPanelFilter.Click
			superGridControl1.PrimaryGrid.FilterExpr = Nothing
		End Sub

		#End Region

		#Region "BtnResetColumnFiltersClick"

		''' <summary>
		''' Handles user request to
		''' reset each column level filter
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnResetColumnFiltersClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnResetColumnFilters.Click
			Dim panel As GridPanel = superGridControl1.PrimaryGrid

			For Each column As GridColumn In panel.Columns
				column.FilterExpr = Nothing
			Next column
		End Sub

		#End Region
	End Class

	#Region "MyCombo"

	''' <summary>
	''' Simple class definition to enable ThreeState
	''' </summary>
	Public Class MyCheckBox
		Inherits GridCheckBoxXEditControl
		Public Sub New()
			ThreeState = True
		End Sub
	End Class

	#End Region

	#Region "MyCombo"

	''' <summary>
	''' Simple class to create combo with
	''' every base employee name added by default
	''' </summary>
	Public Class MyComboBox
		Inherits GridComboBoxExEditControl
		Public Sub New()
			For Each name As String In MyEmployee.LastNames
				If name IsNot Nothing Then
					Items.Add(name)
				Else
					Items.Add("")
				End If
			Next name
		End Sub
	End Class

	#End Region

	#Region "MyEmployee"

	''' <summary>
	''' Simple MyEmployee class definition
	''' </summary>
	Public Class MyEmployee
        Implements INotifyPropertyChanged

		#Region "Private variables"

		Private _FirstName As String
		Private _LastName As String
		Private _Age As Integer
		Private _Id As Integer
		Private _HireDate As Date
        Private _Citizen As Boolean

		Private Shared _Rand As New Random()

        Public Shared FirstNames() As String = {Nothing, "Isabella", "Sophia", "Emma", _
            "Olivia", "Ava", "Emily", "Abigail", "Madison", "Chloe", "Mia", "Addison", _
            "Elizabeth", "Ella", "Natalie", "Samantha", "Alexis", "Lily", "Grace", "Hailey", _
            "Alyssa", "Lillian", "Hannah", "Avery", "Leah", "Nevaeh", "Sofia", "Ashley", _
            "Anna", "Brianna", "Sarah", "Zoe", "Victoria", "Gabriella", "Brooklyn", "Kaylee", _
            "Taylor", "Layla", "Allison", "Evelyn", "Riley", "Amelia", "Khloe"}

        Public Shared LastNames() As String = {Nothing, "Anderson", "Brown", "Clark", "Davis", _
            "García", "González", "Hall", "Harris", "Hernández", "Jackson", "Johnson", "Jones", _
            "Lee", "Lewis", "López", "Martin", "Martínez", "Miller", "Moore", "Pérez", "Robinson", _
            "Rodríguez", "Smith", "Taylor", "Thomas", "Thompson", "Walker", "White", "Williams", _
            "Wilson", "Young"}

		#End Region

		Public Sub New(ByVal firstName As String, ByVal lastName As String, ByVal age As Integer, ByVal id As Integer, ByVal hireDate As Date, ByVal citizen As Boolean)
			_FirstName = firstName
			_LastName = lastName
			_Age = age
			_Id = id
			_HireDate = hireDate
			_Citizen = citizen
		End Sub

		#Region "GetNewEmployee"

		''' <summary>
		''' Gets a new, random employee
		''' </summary>
		''' <returns></returns>
        Public Shared Function GetNewEmployee() As MyEmployee
            Dim fname As String = FirstNames(_Rand.Next(FirstNames.Length))
            Dim lname As String = LastNames(_Rand.Next(LastNames.Length))

            Dim age_Renamed As Integer = _Rand.Next(18, 70)
            Dim id_Renamed As Integer = _Rand.Next(10000, 12000)

            Dim hireDate_Renamed As Date = Date.Now.AddMonths(_Rand.Next(0, 12))

            hireDate_Renamed = hireDate_Renamed.AddYears(_Rand.Next(-20, 20))
            hireDate_Renamed = hireDate_Renamed.AddDays(_Rand.Next(0, 20))

            Return (New MyEmployee(fname, lname, age_Renamed, id_Renamed, _
                    hireDate_Renamed, (age_Renamed + id_Renamed) Mod 2 = 0))
        End Function

		#End Region

		#Region "Public properties"

		#Region "FirstName"

		Public Property FirstName() As String
			Get
				Return (_FirstName)
			End Get

			Set(ByVal value As String)
				_FirstName = value

				OnPropertyChanged("FirstName")
			End Set
		End Property

		#End Region

		#Region "LastName"

		Public Property LastName() As String
			Get
				Return (_LastName)
			End Get

			Set(ByVal value As String)
				_LastName = value

				OnPropertyChanged("LastName")
			End Set
		End Property

		#End Region

		#Region "Age"

		Public Property Age() As Integer
			Get
				Return (_Age)
			End Get

			Set(ByVal value As Integer)
				_Age = value

				OnPropertyChanged("Age")
			End Set
		End Property

		#End Region

		#Region "Id"

		Public Property Id() As Integer
			Get
				Return (_Id)
			End Get

			Set(ByVal value As Integer)
				_Id = value

				OnPropertyChanged("Id")
			End Set
		End Property

		#End Region

		#Region "HireDate"

		Public Property HireDate() As Date
			Get
				Return (_HireDate)
			End Get

			Set(ByVal value As Date)
				_HireDate = value

				OnPropertyChanged("HireDate")
			End Set
		End Property

		#End Region

		#Region "Citizen"

		Public Property Citizen() As Boolean
			Get
				Return (_Citizen)
			End Get

			Set(ByVal value As Boolean)
				_Citizen = value

				OnPropertyChanged("Citizen")
			End Set
		End Property

		#End Region

		#End Region

		#Region "INotifyPropertyChanged Members"

		''' <summary>
		''' Occurs when property value has changed.
		''' </summary>
		Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

		Public Sub OnPropertyChanged(ByVal [property] As String)
			OnPropertyChanged(New PropertyChangedEventArgs([property]))
		End Sub

		''' <summary>
		''' Raises the PropertyChanged event.
		''' </summary>
		''' <param name="e">Event arguments</param>
		Protected Overridable Sub OnPropertyChanged(ByVal e As PropertyChangedEventArgs)
			Dim eh As PropertyChangedEventHandler = PropertyChangedEvent

			If eh IsNot Nothing Then
				eh(Me, e)
			End If
		End Sub

		#End Region
	End Class

	#End Region
End Namespace

File: SuperGrid/C#/DemoBasicFiltering.cs
using System;
using System.ComponentModel;
using DevComponents.DotNetBar;
using DevComponents.DotNetBar.SuperGrid;

namespace SuperGridDemo
{
    public partial class DemoBasicFiltering : Office2007Form
    {
        public DemoBasicFiltering()
        {
            InitializeComponent();

            // Initialize the grid
            // and set the sample description text

            InitializeGrid();

            ShellServices.LoadRtbText(richTextBox1,
                "SuperGridDemo.Resources.DemoBasicFiltering.rtf");
        }

        #region InitializeGrid

        /// <summary>
        /// Initialize the grid
        /// </summary>
        private void InitializeGrid()
        {
            GridPanel panel = superGridControl1.PrimaryGrid;

            panel.Columns["LastName"].EditorType = typeof(MyComboBox);
            panel.Columns["Citizen"].EditorType = typeof(MyCheckBox);

            // Add some initial rows for the user to filter

            AddRows();
        }

        #endregion

        #region AddRows

        /// <summary>
        /// Adds 1000 random data rows for the user to filter
        /// </summary>
        private void AddRows()
        {
            GridPanel panel = superGridControl1.PrimaryGrid;

            superGridControl1.BeginUpdate();

            int count = panel.Rows.Count;

            for (int i = 0; i < 1000; i++)
            {
                MyEmployee emp = MyEmployee.GetNewEmployee();

                object firstName = (emp.Age % 7 == 0 ? (object)DBNull.Value : emp.FirstName);
                object age = (emp.Age % 5 == 0 ? null : (object)emp.Age);
                object citizen = (emp.Age % 3 == 0 ? null : (object)emp.Citizen);
                object hireDate = (emp.Age % 6 == 0 ? null : (object)emp.HireDate);

                object[] ob1 = new object[]
                {
                    count + i, emp.LastName, firstName,
                    age, hireDate, citizen
                };

                if (i < 12)
                {
                    int n = i % 6 + 1;

                    object[] ob2 = new object[n];

                    for (int j = 0; j < n; j++)
                        ob2[j] = ob1[j];

                    panel.Rows.Add(new GridRow(ob2));
                }
                else
                {
                    panel.Rows.Add(new GridRow(ob1));
                }
            }

            superGridControl1.EndUpdate();

            // We'll stop it at 10,000 - just for grins

            if (panel.Rows.Count >= 10000)
                btnAddRows.Enabled = false;
        }

        #endregion

        #region BtnAddRowsClick

        /// <summary>
        /// Handles user request to add another 1000 rows
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnAddRowsClick(object sender, EventArgs e)
        {
            AddRows();
        }

        #endregion

        #region BtnResetPanelFilterClick

        /// <summary>
        /// Handles user request to reset the
        /// row level panel filter
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnResetPanelFilterClick(object sender, EventArgs e)
        {
            superGridControl1.PrimaryGrid.FilterExpr = null;
        }

        #endregion

        #region BtnResetColumnFiltersClick

        /// <summary>
        /// Handles user request to
        /// reset each column level filter
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnResetColumnFiltersClick(object sender, EventArgs e)
        {
            GridPanel panel = superGridControl1.PrimaryGrid;

            foreach (GridColumn column in panel.Columns)
                column.FilterExpr = null;
        }

        #endregion
    }

    #region MyCombo

    /// <summary>
    /// Simple class definition to enable ThreeState
    /// </summary>
    public class MyCheckBox : GridCheckBoxXEditControl
    {
        public MyCheckBox()
        {
            ThreeState = true;
        }
    }

    #endregion

    #region MyCombo

    /// <summary>
    /// Simple class to create combo with
    /// every base employee name added by default
    /// </summary>
    public class MyComboBox : GridComboBoxExEditControl
    {
        public MyComboBox()
        {
            foreach (string name in MyEmployee.LastNames)
                Items.Add(name ?? "");
        }
    }

    #endregion

    #region MyEmployee

    /// <summary>
    /// Simple MyEmployee class definition
    /// </summary>
    public class MyEmployee : INotifyPropertyChanged
    {
        #region Private variables

        private string _FirstName;
        private string _LastName;
        private int _Age;
        private int _Id;
        private DateTime _HireDate;
        private bool _Citizen;

        static Random _Rand = new Random();

        public static string[] FirstNames = new string[] { null,
            "Isabella", "Sophia", "Emma", "Olivia", "Ava", "Emily", "Abigail", "Madison",
            "Chloe", "Mia", "Addison", "Elizabeth", "Ella", "Natalie", "Samantha", "Alexis",
            "Lily", "Grace", "Hailey", "Alyssa", "Lillian", "Hannah", "Avery", "Leah",
            "Nevaeh", "Sofia", "Ashley", "Anna", "Brianna", "Sarah", "Zoe", "Victoria",
            "Gabriella", "Brooklyn", "Kaylee", "Taylor", "Layla", "Allison", "Evelyn",
            "Riley", "Amelia", "Khloe"};

        public static string[] LastNames = new string[] { null,
            "Anderson", "Brown", "Clark", "Davis", "García", "González", "Hall", "Harris",
            "Hernández", "Jackson", "Johnson", "Jones", "Lee", "Lewis", "López", "Martin",
            "Martínez", "Miller", "Moore", "Pérez", "Robinson", "Rodríguez", "Smith",
            "Taylor", "Thomas", "Thompson", "Walker", "White", "Williams", "Wilson", "Young" };

        #endregion

        public MyEmployee(string firstName, string lastName,
            int age, int id, DateTime hireDate, bool citizen)
        {
            _FirstName = firstName;
            _LastName = lastName;
            _Age = age;
            _Id = id;
            _HireDate = hireDate;
            _Citizen = citizen;
        }

        #region GetNewEmployee

        /// <summary>
        /// Gets a new, random employee
        /// </summary>
        /// <returns></returns>
        public static MyEmployee GetNewEmployee()
        {
            string fname = FirstNames[_Rand.Next(FirstNames.Length)];
            string lname = LastNames[_Rand.Next(LastNames.Length)];

            int age = _Rand.Next(18, 70);
            int id = _Rand.Next(10000, 12000);

            DateTime hireDate = DateTime.Now.AddMonths(_Rand.Next(0, 12));
            hireDate = hireDate.AddYears(_Rand.Next(-20, 20));
            hireDate = hireDate.AddDays(_Rand.Next(0, 20));

            return (new MyEmployee(fname, lname, age, id, hireDate, (age + id) % 2 == 0));
        }

        #endregion

        #region Public properties

        #region FirstName

        public string FirstName
        {
            get { return (_FirstName); }

            set
            {
                _FirstName = value;

                OnPropertyChanged("FirstName");
            }
        }

        #endregion

        #region LastName

        public string LastName
        {
            get { return (_LastName); }

            set
            {
                _LastName = value;

                OnPropertyChanged("LastName");
            }
        }

        #endregion

        #region Age

        public int Age
        {
            get { return (_Age); }

            set
            {
                _Age = value;

                OnPropertyChanged("Age");
            }
        }

        #endregion

        #region Id

        public int Id
        {
            get { return (_Id); }

            set
            {
                _Id = value;

                OnPropertyChanged("Id");
            }
        }

        #endregion

        #region HireDate

        public DateTime HireDate
        {
            get { return (_HireDate); }

            set
            {
                _HireDate = value;

                OnPropertyChanged("HireDate");
            }
        }

        #endregion

        #region Citizen

        public bool Citizen
        {
            get { return (_Citizen); }

            set
            {
                _Citizen = value;

                OnPropertyChanged("Citizen");
            }
        }

        #endregion

        #endregion

        #region INotifyPropertyChanged Members

        /// <summary>
        /// Occurs when property value has changed.
        /// </summary>
        public event PropertyChangedEventHandler PropertyChanged;

        public void OnPropertyChanged(string property)
        {
            OnPropertyChanged(new PropertyChangedEventArgs(property));
        }

        /// <summary>
        /// Raises the PropertyChanged event.
        /// </summary>
        /// <param name="e">Event arguments</param>
        protected virtual void OnPropertyChanged(PropertyChangedEventArgs e)
        {
            PropertyChangedEventHandler eh = PropertyChanged;

            if (eh != null)
                eh(this, e);
        }

        #endregion
    }

    #endregion
}